plugin/core/sessions.py

```diff
@@ -191,40 +193,53 @@ def get_capability(self, capability: str) -> 'Optional[Any]':
     def handles_path(self, file_path: 'Optional[str]') -> bool:
         if not file_path:
             return False
-
-        if not self._workspace_folders:
-            return True
-
+        with self.ready_lock:
+            # If we're in a window with no folders, or we're a multi-folder session, then we handle any path.
+            if not self._workspace_folders or self._supports_workspace_folders():
+                return True
+        # We're in a window with folders, and we're a single-folder session.
         for folder in self._workspace_folders:
             if file_path.startswith(folder.path):
                 return True
-
         return False
 
     def update_folders(self, folders: 'List[WorkspaceFolder]') -> None:
-        if self._supports_workspace_folders():
-            added, removed = diff_folders(self._workspace_folders, folders)
-            params = {
-                "event": {
-                    "added": [a.to_lsp() for a in added],
-                    "removed": [r.to_lsp() for r in removed]
+        with self.ready_lock:
+            if self._supports_workspace_folders():
+                added, removed = diff_folders(self._workspace_folders, folders)
+                params = {
+                    "event": {
+                        "added": [a.to_lsp() for a in added],
+                        "removed": [r.to_lsp() for r in removed]
+                    }
                 }
-            }
-            notification = Notification.didChangeWorkspaceFolders(params)
-            self.client.send_notification(notification)
-            self._workspace_folders = folders
+                notification = Notification.didChangeWorkspaceFolders(params)
+                self.client.send_notification(notification)
+                self._workspace_folders = folders
 
     def _initialize(self) -> None:
+        self.ready_lock.acquire()  # released in _handle_initialize_result or _handle_initialize_error
         params = get_initialize_params(self._workspace_folders, self.config)
         self.client.send_request(
             Request.initialize(params),
-            lambda result: self._handle_initialize_result(result))
+            self._handle_initialize_result,
+            self._handle_initialize_error)
 
     def _supports_workspace_folders(self) -> bool:
+        assert self.ready_lock.locked()
         workspace_cap = self.capabilities.get("workspace", {})
         workspace_folder_cap = workspace_cap.get("workspaceFolders", {})
         return workspace_folder_cap.get("supported")
 
+    def supports_workspace_folders(self) -> bool:
```

Comment:
Why do we have two definitions for `supports_workspace_folders` and `_supports_workspace_folders`?
Can we have one?

Reply:
_supports_workspace_folders is the actual implementation, supports_workspace_folders just puts a mutex lock around it. I've renamed _supports_workspace_folders to _unsafe_supports_workspace_folders. It's not supposed to be called outside the class.